presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java

```diff
@@ -1586,7 +1593,172 @@ public static Double stSphericalLength(@SqlType(SPHERICAL_GEOGRAPHY_TYPE_NAME) S
+    @ScalarFunction("ST_Contains")
+    @SqlType(BOOLEAN)
+    public static Boolean stSphericalContains(@SqlType(SPHERICAL_GEOGRAPHY_TYPE_NAME) Slice left, @SqlType(SPHERICAL_GEOGRAPHY_TYPE_NAME) Slice right)
+    {
+        OGCGeometry leftGeometry = deserialize(left);
+        if (leftGeometry.isEmpty()) {
+            return null;
+        }
+
+        OGCGeometry rightGeometry = deserialize(right);
+        if (rightGeometry.isEmpty()) {
+            return null;
+        }
+
+        validateSphericalType("ST_Contains", leftGeometry, EnumSet.of(POLYGON));
+        validateSphericalType("ST_Contains", rightGeometry, EnumSet.of(POINT));
+
+        Polygon polygon = (Polygon) leftGeometry.getEsriGeometry();
+        Point point = (Point) rightGeometry.getEsriGeometry();
+
+        if (point.getY() == MAX_LATITUDE || point.getY() == MIN_LATITUDE) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "When applied to SphericalGeography inputs, ST_Contains only supports points which are not poles.");
+        }
+
+        if (containsEitherPole(polygon)) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "When applied to SphericalGeography inputs, ST_Contains only supports polygons which do not enclose poles.");
+        }
+
+        int k = getNumberIntersectionsWithPolygon(polygon, point);
+        if (k % 2 == 1) {
+            return true;
+        }
+        else {
+            return false;
+        }
+    }
+
+    private static boolean liesInRange(double point, double rangeStart, double rangeEnd)
+    {
+        return point >= rangeStart && point <= rangeEnd;
+    }
+
+    private static boolean edgeCrosses(double edgeStart, double edgeEnd, double longitude)
+    {
+        double start = min(edgeStart, edgeEnd);
+        double end = max(edgeStart, edgeEnd);
+        double distanceVia0 = end - start;
+        double distanceVia180 = start - MIN_LONGITUDE + MAX_LONGITUDE - end;
+        if (distanceVia0 < distanceVia180) {
+            return liesInRange(longitude, start, end);
+        }
+        else {
+            return liesInRange(longitude, MIN_LONGITUDE, start) || liesInRange(longitude, end, MAX_LONGITUDE);
+        }
+    }
+
+    private static boolean edgeCrossesLongitude(double edgeStart, double edgeEnd, double longitude)
+    {
+        double start = min(edgeStart, edgeEnd);
+        double end = max(edgeStart, edgeEnd);
+        double distanceVia0 = end - start;
+        double distanceVia180 = start - MIN_LONGITUDE + MAX_LONGITUDE - end;
+        if (distanceVia0 < distanceVia180) {
+            return liesInRange(longitude, start, end);
+        }
+        else {
+            return liesInRange(longitude, MIN_LONGITUDE, start) || liesInRange(longitude, end, MAX_LONGITUDE);
+        }
+    }
+
+    private static boolean edgeIntersectsTestEdge(Point arcStart, Point arcEnd, Point longitudeStart, Point longitudeEnd)
+    {
+        // y of longstart must be lower than y of longend
```

Comment:
I don't understand how this method works. Would you explain? Why is it important that latitude of the end point is less then latitude of the start point? Why `if (distanceVia0 < distanceVia180) {` condition matters?